spec/requests/sessions_spec.rb

```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+describe 'Sessions', type: :request do
+  let(:recaptcha_id) { 'recaptcha_form' }
+
+  describe '#new' do
+    after { cookies[:login_recaptcha] = nil }
+
+    context 'when recaptcha is not configured' do
+      before do
+        allow(RecaptchaService).to receive(:recaptcha_configured?).and_return(false)
+      end
+
+      it 'does not display recaptcha' do
+        cookies[:login_recaptcha] = true
+        get new_user_session_path
+
+        expect(response.body).to_not include recaptcha_id
+      end
+    end 
+
+    context 'when recaptcha is configured' do
+      before do
+        allow(RecaptchaService).to receive(:recaptcha_configured?).and_return(true)
+      end
+
+      context 'when user has recaptcha cookie' do
+        it 'displays recaptcha' do
+          cookies[:login_recaptcha] = true
+          get new_user_session_path
+
+          expect(response.body).to include recaptcha_id
+        end
+      end
+
+      context 'when recaptcha is specified in query string' do
+        before do
+          get new_user_session_path, params: { recaptcha: true }
+        end
+
+        it 'displays recaptcha' do
+          expect(response.body).to include recaptcha_id
+        end
+
+        it 'sets recaptcha cookie' do
+          expect(cookies[:login_recaptcha]).to eq 'true'
+        end
+      end
+
+      context 'when user is specified in sign in params' do
+        let(:failed_attempts) {}
+        let(:user) { create(:user, failed_attempts: failed_attempts) }
+
+        before { get new_user_session_path, params: { user: { email: user.email } } }
```

Comment:
I feel like the **act** part of [Arrange Act Assert](http://wiki.c2.com/?ArrangeActAssert) might be called before the Arrange looking through the rest of the tests below. I'm wondering if it actually tests the edge case where you setup the attempts or something.

Reply:
This is a good point. Yeah I'm curious if these tests pass without the `before`.

Reply:
Whoops, sorry, I had missed this comment before. To my knowledge, the `before` block will run in whatever context your specs run, so they will use the innermost nested `failed_attempts`/`user` value in these specs. Turns out, I could even delete `let(:failed_attempts) {}` up above and the specs would run just fine even though I don't declare `failed_attempts` until a nested context.